charmcraft/jujuignore.py

```diff
@@ -0,0 +1,235 @@
+            res += re.escape(c)
+    res += r'\Z'
+    return res
+
+
+class _Matcher:
+    """Couple a regex with other metadata for how we should match a given pattern."""
+
+    def __init__(self, line_num: int, orig_rule: str, invert: bool, only_dirs: bool,
+                 regex: typing.Pattern):
+        self.line_num = line_num
+        self.orig_rule = orig_rule
+        self.invert = invert
+        self.only_dirs = only_dirs
+        self.compiled = re.compile(regex, re.DOTALL)
+
+    def match(self, path: str, is_dir: bool) -> str:
+        """Checks if a path matches.
+
+        Returns:
+            Can return one of KEEP, SKIP, FORCEKEEP
+        """
+        if self.only_dirs and not is_dir:
+            return KEEP
+        if self.compiled.match(path):
+            if self.invert:
+                return FORCEKEEP
+            return SKIP
+        return KEEP
+
+
+class JujuIgnore:
+    """Track a set of ignore patterns from a .jujuignore file."""
+
+    def __init__(self, patterns: typing.Iterable[str]):
+        self._matchers = []
+        self._compile_from(patterns)
+
+    def _compile_from(self, patterns):
+        for line_num, rule in enumerate(patterns):
+            # humans like line numbers to start from 1
+            line_num += 1
+            orig_rule = rule
+            rule = rule.lstrip().rstrip('\r\n')
+            if not rule or rule.startswith('#'):
+                continue
+            invert = False
+            if rule.startswith('!'):
+                invert = True
+                rule = rule.lstrip('!')
+            rule = _unescape_rule(rule)
+            only_dirs = False
+            if rule.endswith('/'):
+                only_dirs = True
+                rule = rule.rstrip('/')
+            if not rule.startswith('/'):
+                # A rule that doesn't start with '/' means to match any
+                # subdirectory
+                rule = '**/' + rule
+            regex = _rule_to_regex(rule)
+            m = _Matcher(
+                line_num=line_num,
+                orig_rule=orig_rule,
+                invert=invert,
+                only_dirs=only_dirs,
+                regex=regex,
+            )
+            self._matchers.append(m)
+            logger.debug(
+                'Translated .jujuignore %d "%s" => "%s"',
+                line_num,
+                orig_rule,
+                regex)
+
+    def match(self, path: str, is_dir: bool) -> bool:
+        if not path.startswith('/'):
+            path = '/' + path
+        keep = True
+        for matcher in self._matchers:
+            matchRes = matcher.match(path, is_dir)
+            if matchRes == SKIP:
+                keep = False
+            elif matchRes == FORCEKEEP:
+                keep = True
+                break
+        return not keep
+
+
+# default_juju_ignore is the initial set of ignores.
+# juju itself always includes these before adding the contents of .jujuignore
+default_juju_ignore = '''
```

Comment:
I'm concerned about including PYC files (as we still don't use the `.gitignore` file)

Reply:
These are the defaults from juju. We can diverge, but it means people used to be able to deploy from a directory get different results than deploying a .charm than has been built.
If we don't really intend that you can build in place (which makes me a bit sad) then it matters less to diverge.
Otherwise, that is impetus for them to add a .jujuignore if they don't want the .pyc.

Reply:
It's not a matter of including them or not. 

We will NOT include PYCs, as we plan to support `.gitignore` and PYCs are ignored there. So, at the end, we will not have them included. And now, we don't have them included.

My concern is if including them in the brief period between we "get everything except jujuignored" and "get everything except jujuignored and gitignored" is problematic or not.

Reply:
It's unclear to me whether we actually want to support `.gitignore` over `.jujuignore`. We want to support at least one of them, for sure, and `.jujuignore` does make some sense because it's what `juju deploy .` does for example and the argument that you can have artifacts you want in the charm but not in git, and viceversa, is a reasonable one.

Supporting both is possible, but as I said yesterday in the standup, supporting one and then circling back to that forum post to see what our users thing is a reasonable step.

Having said all that, supporting `.jujuignore` means having this implicit list. We don't get to have opinions about this list, here; if we need it changed we'd have to address it with juju and then update this one to reflect that. We can add `*.py[cod]` to an explicit `.jujuignore`, and in fact #91 does that, and that's probably good enough.